File: csharp/Resources/SettingsResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Environment policy and settings Wave 3 operations
    /// </summary>
    public class SettingsResource
    {
        private readonly AuthdogClient _client;

        public SettingsResource(AuthdogClient client)
        {
            _client = client;
        }

        private Task<JObject> GetAsync(string tenantId, string environmentId, string suffix) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/{suffix}");

        private Task<JObject> PutAsync(string tenantId, string environmentId, string suffix, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Put,
                $"{AuthdogClient.Env(tenantId, environmentId)}/{suffix}",
                body);

        public Task<JObject> GetBotDetectionPolicyAsync(string tenantId, string environmentId) =>
            GetAsync(tenantId, environmentId, "bot-detection-policy");

        public JObject GetBotDetectionPolicy(string tenantId, string environmentId) =>
            GetBotDetectionPolicyAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> UpdateBotDetectionPolicyAsync(string tenantId, string environmentId, object body) =>
            PutAsync(tenantId, environmentId, "bot-detection-policy", body);

        public JObject UpdateBotDetectionPolicy(string tenantId, string environmentId, object body) =>
            UpdateBotDetectionPolicyAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> GetBreachedPasswordPolicyAsync(string tenantId, string environmentId) =>
            GetAsync(tenantId, environmentId, "breached-password-policy");

        public JObject GetBreachedPasswordPolicy(string tenantId, string environmentId) =>
            GetBreachedPasswordPolicyAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> UpdateBreachedPasswordPolicyAsync(string tenantId, string environmentId, object body) =>
            PutAsync(tenantId, environmentId, "breached-password-policy", body);

        public JObject UpdateBreachedPasswordPolicy(string tenantId, string environmentId, object body) =>
            UpdateBreachedPasswordPolicyAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> GetBruteForcePolicyAsync(string tenantId, string environmentId) =>
            GetAsync(tenantId, environmentId, "brute-force-policy");

        public JObject GetBruteForcePolicy(string tenantId, string environmentId) =>
            GetBruteForcePolicyAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> UpdateBruteForcePolicyAsync(string tenantId, string environmentId, object body) =>
            PutAsync(tenantId, environmentId, "brute-force-policy", body);

        public JObject UpdateBruteForcePolicy(string tenantId, string environmentId, object body) =>
            UpdateBruteForcePolicyAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> GetDeviceRiskPolicyAsync(string tenantId, string environmentId) =>
            GetAsync(tenantId, environmentId, "device-risk-policy");

        public JObject GetDeviceRiskPolicy(string tenantId, string environmentId) =>
            GetDeviceRiskPolicyAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> UpdateDeviceRiskPolicyAsync(string tenantId, string environmentId, object body) =>
            PutAsync(tenantId, environmentId, "device-risk-policy", body);

        public JObject UpdateDeviceRiskPolicy(string tenantId, string environmentId, object body) =>
            UpdateDeviceRiskPolicyAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> ListJwtClaimMappingsAsync(string tenantId, string environmentId) =>
            GetAsync(tenantId, environmentId, "jwt-claim-mappings");

        public JObject ListJwtClaimMappings(string tenantId, string environmentId) =>
            ListJwtClaimMappingsAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> SaveJwtClaimMappingAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/jwt-claim-mappings",
                body);

        public JObject SaveJwtClaimMapping(string tenantId, string environmentId, object body) =>
            SaveJwtClaimMappingAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> DeleteJwtClaimMappingAsync(string tenantId, string environmentId, string mappingId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"{AuthdogClient.Env(tenantId, environmentId)}/jwt-claim-mappings/{mappingId}");

        public JObject DeleteJwtClaimMapping(string tenantId, string environmentId, string mappingId) =>
            DeleteJwtClaimMappingAsync(tenantId, environmentId, mappingId).GetAwaiter().GetResult();

        public Task<JObject> GetPasswordPolicyAsync(string tenantId, string environmentId) =>
            GetAsync(tenantId, environmentId, "password-policy");

        public JObject GetPasswordPolicy(string tenantId, string environmentId) =>
            GetPasswordPolicyAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> UpdatePasswordPolicyAsync(string tenantId, string environmentId, object body) =>
            PutAsync(tenantId, environmentId, "password-policy", body);

        public JObject UpdatePasswordPolicy(string tenantId, string environmentId, object body) =>
            UpdatePasswordPolicyAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> GetRateLimitPolicyAsync(string tenantId, string environmentId) =>
            GetAsync(tenantId, environmentId, "rate-limit-policy");

        public JObject GetRateLimitPolicy(string tenantId, string environmentId) =>
            GetRateLimitPolicyAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> UpdateRateLimitPolicyAsync(string tenantId, string environmentId, object body) =>
            PutAsync(tenantId, environmentId, "rate-limit-policy", body);

        public JObject UpdateRateLimitPolicy(string tenantId, string environmentId, object body) =>
            UpdateRateLimitPolicyAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> GetRestrictionsAsync(string tenantId, string environmentId) =>
            GetAsync(tenantId, environmentId, "restrictions");

        public JObject GetRestrictions(string tenantId, string environmentId) =>
            GetRestrictionsAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> UpdateRestrictionsAsync(string tenantId, string environmentId, object body) =>
            PutAsync(tenantId, environmentId, "restrictions", body);

        public JObject UpdateRestrictions(string tenantId, string environmentId, object body) =>
            UpdateRestrictionsAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> GetSessionConfigAsync(string tenantId, string environmentId) =>
            GetAsync(tenantId, environmentId, "session-config");

        public JObject GetSessionConfig(string tenantId, string environmentId) =>
            GetSessionConfigAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> UpdateSessionConfigAsync(string tenantId, string environmentId, object body) =>
            PutAsync(tenantId, environmentId, "session-config", body);

        public JObject UpdateSessionConfig(string tenantId, string environmentId, object body) =>
            UpdateSessionConfigAsync(tenantId, environmentId, body).GetAwaiter().GetResult();
    }
}

File: java/src/main/java/com/authdog/resources/SettingsResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * Environment policy and settings operations.
 */
public final class SettingsResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create a settings helper.
     * @param clientParam Authdog client
     */
    public SettingsResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * Get the bot detection policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode getBotDetectionPolicy(final String tenantId,
                                          final String environmentId)
            throws AuthenticationException, ApiException {
        return get(tenantId, environmentId, "bot-detection-policy");
    }

    /**
     * Update the bot detection policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode updateBotDetectionPolicy(final String tenantId,
                                             final String environmentId,
                                             final Object body)
            throws AuthenticationException, ApiException {
        return put(tenantId, environmentId, "bot-detection-policy",
                body);
    }

    /**
     * Get the breached password policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode getBreachedPasswordPolicy(final String tenantId,
                                              final String environmentId)
            throws AuthenticationException, ApiException {
        return get(tenantId, environmentId,
                "breached-password-policy");
    }

    /**
     * Update the breached password policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode updateBreachedPasswordPolicy(
            final String tenantId,
            final String environmentId,
            final Object body)
            throws AuthenticationException, ApiException {
        return put(tenantId, environmentId,
                "breached-password-policy", body);
    }

    /**
     * Get the brute force policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode getBruteForcePolicy(final String tenantId,
                                        final String environmentId)
            throws AuthenticationException, ApiException {
        return get(tenantId, environmentId, "brute-force-policy");
    }

    /**
     * Update the brute force policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode updateBruteForcePolicy(final String tenantId,
                                           final String environmentId,
                                           final Object body)
            throws AuthenticationException, ApiException {
        return put(tenantId, environmentId, "brute-force-policy",
                body);
    }

    /**
     * Get the device risk policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode getDeviceRiskPolicy(final String tenantId,
                                        final String environmentId)
            throws AuthenticationException, ApiException {
        return get(tenantId, environmentId, "device-risk-policy");
    }

    /**
     * Update the device risk policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode updateDeviceRiskPolicy(final String tenantId,
                                           final String environmentId,
                                           final Object body)
            throws AuthenticationException, ApiException {
        return put(tenantId, environmentId, "device-risk-policy",
                body);
    }

    /**
     * List JWT claim mappings.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return mappings envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode listJwtClaimMappings(final String tenantId,
                                         final String environmentId)
            throws AuthenticationException, ApiException {
        return get(tenantId, environmentId, "jwt-claim-mappings");
    }

    /**
     * Create or update a JWT claim mapping.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return mapping envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode saveJwtClaimMapping(final String tenantId,
                                        final String environmentId,
                                        final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/jwt-claim-mappings",
                body, null, JsonNode.class);
    }

    /**
     * Delete a JWT claim mapping.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param mappingId mapping ID
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode deleteJwtClaimMapping(final String tenantId,
                                          final String environmentId,
                                          final String mappingId)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/jwt-claim-mappings/" + mappingId,
                null, null, JsonNode.class);
    }

    /**
     * Get the password policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode getPasswordPolicy(final String tenantId,
                                      final String environmentId)
            throws AuthenticationException, ApiException {
        return get(tenantId, environmentId, "password-policy");
    }

    /**
     * Update the password policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode updatePasswordPolicy(final String tenantId,
                                         final String environmentId,
                                         final Object body)
            throws AuthenticationException, ApiException {
        return put(tenantId, environmentId, "password-policy", body);
    }

    /**
     * Get the rate limit policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode getRateLimitPolicy(final String tenantId,
                                       final String environmentId)
            throws AuthenticationException, ApiException {
        return get(tenantId, environmentId, "rate-limit-policy");
    }

    /**
     * Update the rate limit policy.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return policy envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode updateRateLimitPolicy(final String tenantId,
                                          final String environmentId,
                                          final Object body)
            throws AuthenticationException, ApiException {
        return put(tenantId, environmentId, "rate-limit-policy", body);
    }

    /**
     * Get environment restrictions.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return restrictions envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode getRestrictions(final String tenantId,
                                    final String environmentId)
            throws AuthenticationException, ApiException {
        return get(tenantId, environmentId, "restrictions");
    }

    /**
     * Update environment restrictions.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return restrictions envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode updateRestrictions(final String tenantId,
                                       final String environmentId,
                                       final Object body)
            throws AuthenticationException, ApiException {
        return put(tenantId, environmentId, "restrictions", body);
    }

    /**
     * Get the session configuration.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return session config envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode getSessionConfig(final String tenantId,
                                     final String environmentId)
            throws AuthenticationException, ApiException {
        return get(tenantId, environmentId, "session-config");
    }

    /**
     * Update the session configuration.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return session config envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode updateSessionConfig(final String tenantId,
                                        final String environmentId,
                                        final Object body)
            throws AuthenticationException, ApiException {
        return put(tenantId, environmentId, "session-config", body);
    }

    /**
     * GET a settings suffix.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param suffix path suffix
     * @return settings envelope
     */
    private JsonNode get(final String tenantId,
                         final String environmentId,
                         final String suffix) {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId) + "/"
                        + suffix,
                null, null, JsonNode.class);
    }

    /**
     * PUT a settings suffix.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param suffix path suffix
     * @param body request body
     * @return settings envelope
     */
    private JsonNode put(final String tenantId,
                         final String environmentId,
                         final String suffix,
                         final Object body) {
        return client.request("PUT",
                EnvPaths.prefix(tenantId, environmentId) + "/"
                        + suffix,
                body, null, JsonNode.class);
    }
}
